File: golangR2/solution/solution0236.go
package solution

func LowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	var lca func(root, p, q *TreeNode) *TreeNode
	lca = func(root, p, q *TreeNode) *TreeNode {
		if root == nil || root.Val == p.Val || root.Val == q.Val {
			return root
		}
		root.Left = lca(root.Left, p, q)
		root.Right = lca(root.Right, p, q)

		//如果左边也能找到相等的，右边也能找到相等的, 则表示当前节点就是最近公共祖先
		if root.Left != nil && root.Right != nil {
			return root
		}

		if root.Left != nil {
			return root.Left
		} else {
			return root.Right
		}
	}

	return lca(root, p, q)
}

File: JavaR3/src/solution/Solution0236.java
package solution;

public class Solution0236 {
	public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
		/**
		 * node是最近公共祖先有3种情况
		 * 1、p，q分别位于root 左右两侧
		 * 2、p=root, q位于root左子树或右子树
		 * 3、q=root, p位于root左子树或右子树
		 */

		if (root == null) {
			return null;
		} else {

			if (root.val == p.val || root.val == q.val) {
				return root;
			}

			TreeNode leftNode = lowestCommonAncestor(root.left, p, q);
			TreeNode rightNode = lowestCommonAncestor(root.right, p, q);

			//leftNode和rightNode都为空，则没有找到
			if (leftNode == null && rightNode == null) {
				return null;
			}

			//p,q分别位于root两侧
			if (leftNode != null && rightNode != null) {
				return root;
			}

			//leftNode和rightNode其中一个为空，那么p,q都在左或者右子树,那么最近公共祖先就是子树的第一个节点
			if (leftNode != null) {
				return leftNode;
			}
            return rightNode;
        }

	}

}
